update infrastructure.rst/compilers (#1950)
conda-forge/conda-forge.github.io: src/maintainer/infrastructure.rst

Conda-forge builds and maintains its own set of compilers for various languages
and/or systems (e.g., ``C``, ``FORTRAN``, ``C++``, ``CUDA``, etc.). These are used
in all of our CI builds to build essentially all artefacts published by conda-forge.

This compiler infrastructure has a critical role beyond building everything, which
is to ensure that packages stay compatible with each other. This is due to how compiled

> packages have a so-called `Application Binary Interface <https://pypackaging-native.github.io/background/binary_interface>`_

Reviewer: ```suggestion
packages have a so-called `Application Binary Interface <https://pypackaging-native.github.io/background/binary_interface>`__
```

These compiler generations may or may not have some unofficial names for our
internal use (e.g. ``comp7``). We note again that the existence of these names
does not imply any level of support or stability for the compilers
that form the given stack.

For the cases that do not require a complete rebuild of conda-forge (i.e. if the ABI
of a new compiler remains compatible), we can just increase the version in our global
pinning, and it will slowly roll out to the ecosystem as feedstocks get rerendered.
For such ABI-compatible upgrades, similar but looser principles apply:

* The pins are similarly in the global pinning, see :ref:`globally_pinned_packages`.
* We provide no support of any kind in terms of the long-term availability of a given compiler version.
* We generally provide notice in the form of an announcement when a compiler is going to be upgraded.
* Without promising any timelines, our compilers on Linux and OSX are normally
  very recent; on Windows, we generally use the last supported VS version.

Despite the lack of explicit support, we try to keep the compilers in their various versions
working also outside of conda-forge, and even provide an easy way to install them

> (through the `compilers <https://github.com/conda-forge/compilers-feedstock>`_ feedstock).

Reviewer: ```suggestion
(through the `compilers <https://github.com/conda-forge/compilers-feedstock>`__ feedstock).
```

* [CUDA] https://github.com/conda-forge/nvcc-feedstock (CUDA infra currently being overhauled)
* [Rust] `Activation <https://github.com/conda-forge/rust-activation-feedstock>`__
  and `Implementation <https://github.com/conda-forge/rust-feedstock>`__
* [Go] `Activation <https://github.com/conda-forge/go-activation-feedstock>`__
  and `Implementation <https://github.com/conda-forge/go-feedstock>`__

To upgrade the compiler version of our default compilers in the global pinning for
Linux or OSX, ensure that the respective above-mentioned feedstocks have been rebuilt
for the new major version, that all interrelated versions are lifted at the same time,
and obviously that the compilers work (e.g. by testing them on some feedstocks by
specifying the new version through the feedstock-local `conda_build_config.yaml`).
You should also check the compiler release notes for warnings about ABI incompatibilities,
and mention any such notices in the discussion about the upgrade.

For Windows, we stay on older compilers for longer, because using a newer toolchain
would force everyone wanting to locally develop with conda-forge artefacts to use

> a toolchain that's at least as new. A more in-depth explanation can be found
> `here <https://github.com/conda-forge/conda-forge.github.io/issues/1732>`_.

Reviewer: It's usually better to have a link text that requires no context to understand its meaning (see [this accessibility guidelines section](https://usability.yale.edu/web-accessibility/articles/links#link-text) for details).

So this could be rephrased like:

```suggestion
a toolchain that's at least as new. You can find more details about this topic in this
`issue about updating to the vc142 toolchain <https://github.com/conda-forge/conda-forge.github.io/issues/1732>`_.
```

(or something similar).
